types/utils/either.ts

```diff
@@ -0,0 +1,3 @@
+type Only<T, U> = { [P in keyof T]: T[P] } & { [P in keyof U]?: never };
+
+export type Either<T, U> = Only<T, U> | Only<U, T>;
```

Comment:
I like that there's an error, although I'm not in love with the message - it's a bit confusing. If you haven't already, can you check for:
1. Any open TypeScript issues where they discuss this. I'm curious if they have plans to change the behavior of the union type or why they haven't already
2. Any eslint rules a customer could use to cause the error themselves

Also, can you confirm that the messages for the existing errors remain the same and haven't regressed?

Reply:
Okay, I tried to find any issues that discuss this behavior. From everything I could find, this seems to be the intended behavior of union types. I would not expect any changes there.

ESLint does not have any particular rules to cover this, but typescript-eslint came close. There were 4 rules there involving union types:

1. Disallow duplicate constituents of union or intersection types
2. Disallow members of unions and intersections that do nothing or override type information
3. Enforce constituents of a type union/intersection to be sorted alphabetically
4. Disallow two overloads that could be unified into one with a union or an optional/rest parameter

None of those rules specifically address this scenario however.

I think you are right about the overall goal. Even though it is preferable to prevent users from doing the wrong thing, we also want them to understand why it is the wrong thing.

The error messages produced by these new types are very confusing. For example, the modified `Only` type gives me this error message when I use `topicEndpointUri` and `connection`:

```
Argument of type '{ topicEndpointUri: string; connection: string; }' is not assignable to parameter of type 'EventGridOutputOptions'.
  Type '{ topicEndpointUri: string; connection: string; }' is not assignable to type 'Only<EventGridOutputConnectionOptions, EventGridOutputKeyOptions>'.
    Type '{ topicEndpointUri: string; connection: string; }' is not assignable to type 'Omit<{ topicEndpointUri?: never; topicKeySetting?: never; }, "connection">'.
      Types of property 'topicEndpointUri' are incompatible.
        Type 'string' is not assignable to type 'never'.ts(2345)
```

That is truly awful. 

It's much clearer to keep the comment we have now that warns users. If they hit problems, I expect most will see the comment and that will take care of things.

I can abandon this PR and we can revisit if it ever becomes a recurring issue for users.